ssz/utils.py

```diff
@@ -1,67 +1,120 @@
+        raise ValueError("Item size must be positive integer")
+    elif CHUNK_SIZE % item_size != 0:
+        raise ValueError("Item size must be a divisor of chunk size")
+    elif item_size <= CHUNK_SIZE:
+        return CHUNK_SIZE // item_size
     else:
-        try:
-            element_sedes = infer_sedes(value[0])
-        except TypeError:
-            raise TypeError("Could not infer sedes for list elements")
-        else:
-            return List(element_sedes)
+        raise Exception("Invariant: unreachable")
 
 
-def infer_sedes(value):
-    """
-    Try to find a sedes objects suitable for a given Python object.
-    """
-    if isinstance(value.__class__, BaseSedes):
-        # Mainly used for `Serializable` Classes
-        return value.__class__
+def pack(serialized_values: Sequence[bytes]) -> Tuple[bytes]:
+    if len(serialized_values) == 0:
+        return (EMPTY_CHUNK,)
 
-    elif isinstance(value, bool):
-        return boolean
+    item_size = len(serialized_values[0])
+    items_per_chunk = get_items_per_chunk(item_size)
 
-    elif isinstance(value, int):
-        raise TypeError("uint sedes object or uint string needs to be specified for ints")
+    number_of_items = len(serialized_values)
+    number_of_chunks = (number_of_items + (items_per_chunk - 1)) // items_per_chunk
 
-    elif isinstance(value, (bytes, bytearray)):
-        return bytes_sedes
+    chunk_partitions = partition(items_per_chunk, serialized_values, pad=b"")
+    chunks_unpadded = (b"".join(chunk_partition) for chunk_partition in chunk_partitions)
 
-    elif isinstance(value, Sequence):
-        return infer_list_sedes(value)
+    full_chunks = tuple(take(number_of_chunks - 1, chunks_unpadded))
+    last_chunk = first(chunks_unpadded)
+    if len(tuple(chunks_unpadded)) > 0:
+        raise Exception("Invariant: all chunks have been taken")
 
-    elif isinstance(value, Iterable):
-        raise TypeError("Cannot infer list sedes for iterables that are not sequences")
+    return full_chunks + (last_chunk.ljust(CHUNK_SIZE, b"\x00"),)
 
-    else:
-        raise TypeError(f"Did not find sedes handling type {type(value).__name__}")
 
+def get_next_power_of_two(integer: int) -> int:
+    powers_of_two = (2**exponent for exponent in count(0))
+    greater_or_equal_powers = dropwhile(lambda power: power < integer, powers_of_two)
+    return first(greater_or_equal_powers)
 
-def get_duplicates(values):
-    counts = collections.Counter(values)
-    return tuple(
-        item
-        for item, num in counts.items()
-        if num > 1
+
+def pad_chunks(chunks: Sequence[bytes]) -> Tuple[bytes]:
+    unpadded_number_of_chunks = len(chunks)
+    padded_number_of_chunks = get_next_power_of_two(unpadded_number_of_chunks)
+    padding = (EMPTY_CHUNK,) * (padded_number_of_chunks - unpadded_number_of_chunks)
+    return tuple(chunks) + padding
+
+
+def hash_layer(child_layer: Sequence[bytes]) -> Tuple[bytes]:
+    child_pairs = partition(2, child_layer)
```

Comment:
I *think* you know that `partition` drops *extra* elements but reading this code it's not clear whether this has been taken care of above this calling level.

Said differently, if callers of `hash_layer` haven't not properly ensured that `len(child_layer)` is even then this will blindly drop the odd trailing element.

I think that we should either 1) add that validation inline or 2) if we *know* that we can't afford the performance hit, then add a comment here explaining that callers of this function are responsible for properly normalizing inputs and that if they do not then bad things will happen.